src/plugins/data_view_management/public/management_app/data_view_management_service.ts

```diff
@@ -0,0 +1,293 @@
+    value: string;
+    name: string;
+    checked?: FilterChecked;
+  }>;
+  indexedFieldTypes: string[];
+  fieldConflictCount: number;
+  tags: Array<{ key: string; 'data-test-subj': string; name: string }>;
+  isRefreshing: boolean;
+  defaultIndex: string;
+  conflictFieldsUrl: string;
+}
+
+const defaultDataViewEditorState: DataViewMgmtState = {
+  allowedTypes: [],
+  relationships: [],
+  fields: [],
+  scriptedFields: [],
+  scriptedFieldLangs: [],
+  indexedFieldTypes: [],
+  fieldConflictCount: 0,
+  tags: [],
+  isRefreshing: true,
+  defaultIndex: '',
+  conflictFieldsUrl: '',
+};
+
+export const stateSelectorFactory =
+  <S>(state$: Observable<S>) =>
+  <R>(selector: (state: S) => R, equalityFn?: (arg0: R, arg1: R) => boolean) =>
+    state$.pipe(map(selector), distinctUntilChanged(equalityFn));
+
+export class DataViewMgmtService {
+  constructor({
+    services: { dataViews, savedObjectsManagement, uiSettings, application },
+    initialValues: {},
+  }: DataViewMgmtServiceConstructorArgs) {
+    this.services = {
+      application,
+      dataViews,
+      savedObjectsManagement,
+      uiSettings,
+    };
+
+    this.internalState$ = new BehaviorSubject<DataViewMgmtState>({
+      ...defaultDataViewEditorState,
+    });
+
+    this.state$ = this.internalState$ as BehaviorObservable<DataViewMgmtState>;
+
+    // allowed types are set once and never change
+    this.allowedTypes = new Promise((resolve) => {
+      savedObjectsManagement.getAllowedTypes().then((resp) => {
+        resolve(resp);
+      });
+    });
+  }
+
+  private services: {
+    application: ApplicationStart;
+    dataViews: DataViewsPublicPluginStart;
+    savedObjectsManagement: SavedObjectsManagementPluginStart;
+    uiSettings: IUiSettingsClient;
+  };
+
+  private allowedTypes: Promise<SavedObjectManagementTypeInfo[]>;
+
+  private internalState$: BehaviorSubject<DataViewMgmtState>;
+  state$: BehaviorObservable<DataViewMgmtState>;
+
+  private updateState = (newState: Partial<DataViewMgmtState>) => {
+    this.internalState$.next({ ...this.state$.getValue(), ...newState });
+  };
+
+  private getKbnUrl = (dataViewId: string) =>
+    setStateToKbnUrl(
+      APP_STATE_STORAGE_KEY,
+      {
+        fieldTypes: ['conflict'],
+        tab: 'indexedFields',
+      },
+      { useHash: this.services.uiSettings.get('state:storeInSessionStorage') },
+      this.services.application.getUrlForApp('management', {
+        path: `/kibana/dataViews/dataView/${encodeURIComponent(dataViewId)}`,
+      })
+    );
+
+  private getTags = async (dataView: DataView) => {
+    if (dataView) {
+      const defaultIndex = await this.services.uiSettings.get('defaultIndex');
+      const tags = getTags(
+        dataView,
+        dataView.id === defaultIndex,
+        this.services.dataViews.getRollupsEnabled()
+      );
+
+      return tags;
+    }
+    return [];
+  };
+
+  async updateScriptedFields() {
+    const dataView = this.state$.getValue().dataView;
```

Comment:
We could pass `dataView` as a parameter here (or even `scriptedFields`) to make sure that it's the latest one as in `setDataView`.

Another option would be to convert it into `getScriptedFieldLangs(scriptedFields)` and call it directly when updating state in `setDataView`

Reply:
I don't want to change a dataView here because I want each of the methods on this service to make atomic updates to state. There should _always_ be a dataView selected when this is called - I could throw if there isn't a dataView selected. 

Usually for a service like this it 1:1 matches the entity being loaded (such as a data view) - perhaps I should do that? It does load the allowedTypes only once. 

I think this might be hitting on a weakness of this method of creating a state service that I've seen better done with XState and perhaps other libraries - sometimes you want a specific state to have a limited set of next states. Calling this method should only be possible when a dataView is selected. I guess I should be able to change the shape of my service to make this possible.

At this very moment I'm not sure which portion this is worth addressing.

Reply:
I think I'd prefer to leave this alone for now, or address it as a follow up. 

I think `setDataView` could return the bulk of the api instead of having it off the main service. I think it would be a good demonstration of a technique but its not necessary for the completion of this PR.